lib/decoders/layer.js

```diff
@@ -115,18 +109,86 @@ class Layer extends Duplex {
+   * a soqlRow which does not fit in the set of layers that we know about.
+   * _updateColumnTypes mutates the Layer's column types as values that
+   * have more type info are written to it.
+   *
+   * An Example:
+   *   A dataset is being read which looks like
+   *   [
+   *     ['a', 1, POINT(1, 2), null],
+   *     ['b', 1, POINT(2, 3), 'something']
+   *     ['c', 1, POINT(3, 4), null]
+   *   ]
+   *
+   *
+   *   When we encounter the first row, a Layer, layer0, is created with the first soqlRow:
+   *     ['a', 1, POINT(1, 2), null]
+   *   So we end up with layer0 that has columns of type
+   *     [SoQLText, SoQLNumber, SoQLPoint, SoQLNull]
+   *
+   *   Then we encounter the second row, row1, which has more info about what the last
+   *   column actually is.
+   *     ['b', 1, POINT(2, 3), 'something']
+   *   layer0.belongsIn(row1) will return true, so row1 will get written to layer0.
+   *   layer0.write will call _updateColumnTypes to update its undefined columns to reflect
+   *   the fact that it is now storing values with more type info than it had before.
+   *
+   * This is a little confusing and gross, and comes from the fact that this is
+   * being written to the lowest common denominator (geoJSON, KML, KMZ) of type
+   * info. ESRI shapefiles give more hints about types, though all those hints
+   * are eliminated in the shapefile library being used.
+   *
+   * This could be written as a big map, but it is called on every write,
+   * so it's important to return as early as possible.
+   *
+   * @mutates this.columns
+   */
+  _updateColumnTypes(soqlRow) {
+    var undefinedColumns = this.columns.filter((c) => c.ctype === 'null');
+    if (!undefinedColumns.length) return;
+
+    var definedSoqlCols = soqlRow.filter((c) => c.ctype !== 'null');
+    //We want the set of columns that are defined for the value, but are not defined
+    //for the layer. These are the columns that we will replace in the layer.
+    var toAdd = definedSoqlCols.filter((valCol) => {
+      return undefinedColumns.find((col) => valCol.rawName === col.rawName);
+    });
+
+    if(toAdd.length === 0) return;
+
+    this.columns = this.columns.map((col) => {
+      var valCol = toAdd.find((newCol) => newCol.rawName === col.rawName);
+      var newCol;
+      if(valCol) {
+        newCol = new valCol.constructor(valCol.rawName);
+        logger.debug(`Replacing old undefined column ${valCol.rawName} with new ${valCol.constructor.name} column`);
+      }
+      return newCol || col;
+    });
+  }
 
+  /**
+   * Write the soqlRow with a given crs to the layer.
+   * This will only be called if belongsIn returned true for the soqlRow and this layer.
+   * See the note above on the confusing mutate-y  bits.
+   * @param  {[type]} crs       [description]
+   * @param  {[type]} soqlRow   [description]
+   * @param  {[type]} throwaway [description]
+   * @return {[type]}           [description]
+   */
```

Comment:
the documentation so far on this pr has been so, so good it's a little disappointing to not get a description of what crs is!